icicle/appUtils/msm/msm.cu

```diff
@@ -450,74 +441,105 @@ namespace msm {
-        offsets_temp_storage, offsets_temp_storage_bytes, bucket_sizes, bucket_offsets, total_nof_buckets, stream));
+        offsets_temp_storage, offsets_temp_storage_bytes, bucket_sizes, bucket_offsets, total_nof_buckets + 1, stream));
       CHK_IF_RETURN(cudaMallocAsync(&offsets_temp_storage, offsets_temp_storage_bytes, stream));
       CHK_IF_RETURN(cub::DeviceScan::ExclusiveSum(
-        offsets_temp_storage, offsets_temp_storage_bytes, bucket_sizes, bucket_offsets, total_nof_buckets, stream));
+        offsets_temp_storage, offsets_temp_storage_bytes, bucket_sizes, bucket_offsets, total_nof_buckets + 1, stream));
       CHK_IF_RETURN(cudaFreeAsync(offsets_temp_storage, stream));
 
+      A* d_points;
+      cudaStream_t stream_points;
+      if (!are_points_on_device || are_points_montgomery_form) CHK_IF_RETURN(cudaStreamCreate(&stream_points));
+      if (!are_points_on_device) {
+        // copy points to gpu
+        CHK_IF_RETURN(cudaMallocAsync(&d_points, sizeof(A) * nof_points, stream_points));
+        CHK_IF_RETURN(cudaMemcpyAsync(d_points, points, sizeof(A) * nof_points, cudaMemcpyHostToDevice, stream_points));
+      } else {
+        d_points = points;
+      }
+
+      if (are_points_montgomery_form) {
+        if (are_points_on_device) {
+          A* d_mont_points;
+          CHK_IF_RETURN(cudaMallocAsync(&d_mont_points, sizeof(A) * nof_points, stream_points));
+          CHK_IF_RETURN(mont::FromMontgomery(d_points, nof_points, stream_points, d_mont_points));
+          d_points = d_mont_points;
+        } else
+          CHK_IF_RETURN(mont::FromMontgomery(d_points, nof_points, stream_points, d_points));
+      }
+      cudaEvent_t event_points_uploaded;
+      if (!are_points_on_device || are_points_montgomery_form) {
+        CHK_IF_RETURN(cudaEventCreateWithFlags(&event_points_uploaded, cudaEventDisableTiming));
+        CHK_IF_RETURN(cudaEventRecord(event_points_uploaded, stream_points));
+      }
+
+      P* buckets;
+      CHK_IF_RETURN(cudaMallocAsync(&buckets, sizeof(P) * (total_nof_buckets + nof_bms_in_batch), stream));
+
+      // launch the bucket initialization kernel with maximum threads
+      NUM_THREADS = 1 << 10;
+      NUM_BLOCKS = (total_nof_buckets + nof_bms_in_batch + NUM_THREADS - 1) / NUM_THREADS;
+      initialize_buckets_kernel<<<NUM_BLOCKS, NUM_THREADS, 0, stream>>>(buckets, total_nof_buckets + nof_bms_in_batch);
+
+      // removing zero bucket, if it exists
+      unsigned smallest_bucket_index;
+      CHK_IF_RETURN(cudaMemcpyAsync(
+        &smallest_bucket_index, single_bucket_indices, sizeof(unsigned), cudaMemcpyDeviceToHost, stream));
+      unsigned zero_bucket_offset = (smallest_bucket_index == 0) ? 1 : 0;
+
       // sort by bucket sizes
       unsigned h_nof_buckets_to_compute;
       CHK_IF_RETURN(cudaMemcpyAsync(
         &h_nof_buckets_to_compute, nof_buckets_to_compute, sizeof(unsigned), cudaMemcpyDeviceToHost, stream));
-
-      // if all points are 0 just return point 0
```

Comment:
There was a question about removing this block. Turned out it was there for a good reason for this block to be here. If all the scalars are zero, there's a division by zero error inside. I will push a more elegant fix for this tomorrow.